File: 03_tuples_arrays/tuples.py
from collections import namedtuple

def reverse(pair: tuple) -> tuple:
    a_int, b_bool = pair
    return (b_bool, a_int)

Matrix = namedtuple('Matrix', ['x0', 'x1', 'x2', 'x3'])

if __name__ == "__main__":
    many_types = (1,2,'a',True)

    print("first value: ", many_types[0])
    print("second value: ", many_types[1])

    tuple_of_tuples = ((1,2,2), (4,-1,-2))
    print("tuple of tuples: ", tuple_of_tuples)

    pair = (1, True)
    print("pair is ", pair)
    print("the reverse pair is ", reverse(pair))

    print("one element tuple: {:?}", (5, ))
    print("just one integer: {:?}", (5))

    _tuple = (1, "hello", 4.5, True)
    a, b, c, d, = _tuple

    matrix = Matrix(1.1, 1.2, 2.1, 2.2)
    print(matrix)

File: 03_tuples_arrays/tuples.rs
fn reverse(pair: (i32, bool)) -> (bool, i32) {
    let (a_int, b_bool) = pair;
    (b_bool, a_int)
}

#[derive(Debug)]
struct Matrix(f32, f32, f32, f32);

fn main() {
    let many_types = (1u8, 2u16, 'a', true);
    println!("first value: {}", many_types.0);
    println!("second value: {}", many_types.1);

    let tuples_of_tuples = ((1u8, 2u16, 2u32), (4u64, -1i8), -2i16);
    println!("tuple of tuples: {:?}", tuples_of_tuples);

    let pair = (1, true);
    println!("pair is {:?}", pair);
    println!("the reverse pais is {:?}", reverse(pair));

    println!("one element tuple: {:?}", (5u32,)); // note extra comma
    println!("just an integer: {}", (5u32));

    let tuple = (1, "hello", 4.5, true);
    let (a, b, c, d) = tuple;
    println!("{:?}, {:?}, {:?}, {:?}", a, b, c, d);

    let matrix = Matrix(1.2, 2.1, 2.3, 4.1);
    println!("{:?}", matrix)
}
